server/sonar-server/src/main/java/org/sonar/server/es/IndexerStartupTask.java

```diff
@@ -54,16 +62,55 @@ private boolean indexesAreEnabled() {
   }
 
   private void indexEmptyTypes(StartupIndexer indexer) {
-    Set<IndexType> emptyTypes = getEmptyTypes(indexer);
-    if (!emptyTypes.isEmpty()) {
-      log(indexer, emptyTypes, "...");
-      indexer.indexOnStartup(emptyTypes);
-      log(indexer, emptyTypes, "done");
+    Set<IndexType> uninizializedTypes = getUninitializedTypes(indexer);
+    if (!uninizializedTypes.isEmpty()) {
+      log(indexer, uninizializedTypes, "...");
+      indexer.indexOnStartup(uninizializedTypes);
+      uninizializedTypes.forEach(this::setInitialized);
+      log(indexer, uninizializedTypes, "done");
     }
   }
 
-  private Set<IndexType> getEmptyTypes(StartupIndexer indexer) {
-    return indexer.getIndexTypes().stream().filter(esClient::isEmpty).collect(toSet());
+  private Set<IndexType> getUninitializedTypes(StartupIndexer indexer) {
+    return indexer.getIndexTypes().stream().filter(this::getUninitialized).collect(toSet());
+  }
+
+  private boolean getUninitialized(IndexType indexType) {
+    return esClient.nativeClient().admin().indices().prepareGetSettings(indexType.getIndex()).get().getSetting(indexType.getIndex(),
+      getInitializedName(indexType)) == null;
+  }
+
+  private void setInitialized(IndexType indexType) {
+    String index = indexType.getIndex();
+    closeIndex(index);
+    setIndexSetting(index, getInitializedName(indexType), true);
+    openIndex(index);
+    waitForIndexGreen(index);
+  }
+
+  private void closeIndex(String index) {
+    Client nativeClient = esClient.nativeClient();
+    CloseIndexAction.INSTANCE.newRequestBuilder(nativeClient).setIndices(index).get();
+  }
+
+  private void setIndexSetting(String index, String name, boolean value) {
+    Client nativeClient = esClient.nativeClient();
+    Builder setting = org.elasticsearch.common.settings.Settings.builder().put(name, value);
+    nativeClient.admin().indices().prepareUpdateSettings(index).setSettings(setting).get();
+  }
+
+  private void openIndex(String index) {
+    Client nativeClient = esClient.nativeClient();
+    OpenIndexAction.INSTANCE.newRequestBuilder(nativeClient).setIndices(index).get();
+  }
+
+  private void waitForIndexGreen(String index) {
+    Client nativeClient = esClient.nativeClient();
+    ClusterHealthAction.INSTANCE.newRequestBuilder(nativeClient).setIndices(index).setWaitForStatus(ClusterHealthStatus.GREEN).get(TimeValue.timeValueMinutes(10));
```

Comment:
Status can be yellow if replica are not fully up. That can occur in cluster mode. So the correct statement is `setWaitForYellowStatus()`.